Feature/cele 32

afonsobspinto opened the pull request:
Closes first iteration of https://metacell.atlassian.net/browse/CELE-32

- Connect start from template button
- Render graph with cytoscape and connect it to React context
- Adds legend and menu prototypes
- Adds different layout options
- Adds different coloring options
- Adds show connected option and show connected as individual cells
- Removes placeholder left and right components
- Changes workspace model to store a record of neuronID-> neuron for all the available neurons of the workspace (this includes all neurons existent in all the active datasets). This is required because /api/connections doesn't provide all the information necessary for the 'connected' neurons. The original nemanode code fetches all neuron information from all the datasets in the beginning of the application.

- Adds "/api/connections" endpoint (not yet equivalent to the original nemanode endpoint due to different threshold strategiess)
- 


On review please have a look at the following questions:
https://metacell.atlassian.net/browse/CELE-32?focusedCommentId=14503

https://github.com/MetaCell/c-elegans-app/assets/19196034/3c39c3cc-ad27-46d4-a1a4-3c8290b7e783

afonsobspinto committed:
CELE-32 feat: Add basic 2D viewer

afonsobspinto committed:
CELE-32 feat: Change to light theme

afonsobspinto committed:
CELE-32 chore: Fix build errors

afonsobspinto committed:
CELE-32 feat: Connect start from template

afonsobspinto committed:
CELE-32 fix: Add light theme

afonsobspinto committed:
CELE-32 feat: Update workspaces to include content instead of just ids

afonsobspinto committed:
CELE-32 feat: Connect 2D Viewer with context

afonsobspinto committed:
CELE-32 chore: Remove placeholder components

afonsobspinto committed:
CELE-32 feat: Add connectivity endpoint

afonsobspinto committed:
CELE-32 feat: Connect twoDViewer with new endpoint

afonsobspinto committed:
CELE-32 refactor: Refactor TwoDViewer component

afonsobspinto committed:
CELE-32 feat: Add resize observer

afonsobspinto committed:
CELE-32 feat: Add 2D menu prototype

afonsobspinto committed:
CELE-32 feat: Add layouts

afonsobspinto committed:
CELE-32 feat: Add 2D legend

afonsobspinto committed:
CELE-32 feat: Add color map strategy menu

afonsobspinto committed:
CELE-32 feat: Add color map strategy pattern

afonsobspinto committed:
CELE-32 feat: Connect coloring strategy

afonsobspinto committed:
CELE-32 feat: WIP - Allow multiple colors

afonsobspinto committed:
CELE-32 feat: WIP - Add twoD menu switches

afonsobspinto committed:
CELE-32 feat: Update workspace properties

afonsobspinto committed:
CELE-32 feat: Add show connected as individual cells

afonsobspinto committed:
CELE-32 feat: Add threshold buttons

afonsobspinto committed:
CELE-32 chore: Apply backend linting

afonsobspinto committed:
CELE-32 chore: Fix frontend build

afonsobspinto committed:
CELE-32 chore: Fix frontend build

afonsobspinto committed:
CELE-32 chore: Add missing files

afonsobspinto committed:
CELE-32 chore: Merge with develop

aranega commented on applications/visualizer/backend/api/schemas.py:
```diff
@@ -51,9 +53,12 @@ class Meta:
 
 
 class Connection(ModelSchema, BilingualSchema):
+    annotations: List[str] = []
```
No need to use `List`, or `Dict` or to import `typing`, you can use directly `list` and `dict` since Python 3.10 (I'm not sure about the version), but as we are using Python 3.11, we can use it :)

aranega commented on applications/visualizer/backend/api/services/connectivity.py:
```diff
@@ -0,0 +1,122 @@
+    cells, dataset_type, include_annotations, include_neighboring_cells
+):
+    annotations_map = {}
+    if include_annotations:
```
It's a small detail (and you don't have to change it if you don't feel like it), but this is a perfect candidate here were having the negative condition and returning directly gains in clarity and avoid nested code:

```python
def _query_annotation(cells, ...):
  if not include_annotations:
    return {}  # We can se directly we return an empty map if we don't include the annotation
  annotations_map = {}
  annotations = Annotation.objects.filter( ... )
  # [...] rest of the code
  return annotations_map
```

aranega commented on applications/visualizer/backend/api/services/connectivity.py:
```diff
@@ -0,0 +1,122 @@
+        )
+        for annotation in annotations:
+            key = _get_connection_key(annotation.pre, annotation.post, annotation.type)
+            if key in annotations_map:
```
I think you can use a `defaultlist` as you did elswehere in the code. If you don't want to have the import here for some reasons, you can use `setdefault` which will be faster as it will avoid the check of the presence of the key and the addition later.

```python
annotations_map.setdefault(key, []).append(annotation.annotation)
```

aranega commented on applications/visualizer/frontend/src/helpers/workspaceHelper.ts:
```diff
@@ -0,0 +1,14 @@
+import {Dataset, DatasetsService} from "../rest";
+
+export async function fetchDatasets(datasetIds: Set<string>): Promise<Record<string, Dataset>> {
```
It would be better to have a endpoint in the backend that would fetch a set of datasets following ids. That would reduce the number of calls and avoid the use of the promise map, simplifying the code.

aranega commented on applications/visualizer/frontend/src/models/models.ts:
```diff
@@ -23,3 +22,7 @@ export interface NeuronGroup {
     neurons: Set<string>;
 }
+
+export interface ColorMap {
```
Is this interface a leftover from a previous version of the color strategy?

aranega commented on applications/visualizer/frontend/src/components/viewers/TwoD/TwoDViewer.tsx:
```diff
@@ -0,0 +1,188 @@
+    };
+
+    const updateNodeColors = () => {
+        if (cyRef.current) {
```
Here also, this is a matter of taste, but writing the negation of the condition with a return shows quickly that the only behavior implemented is when there is a `.current` value (and avoids the nested code)

aranega commented on applications/visualizer/frontend/src/components/viewers/TwoD/TwoDViewer.tsx:
```diff
@@ -0,0 +1,188 @@
+        const edges = [];
+
+        connections.forEach(conn => {
+            nodes.add(conn.pre);
```
Just a quick info about `.add`, you chan chain them :D 

```js
nodes.add(conn.pre).add(conn.post)
```

aranega commented on applications/visualizer/frontend/src/helpers/twoD/coloringStrategy/CellTypeColoringStrategy.ts:
```diff
@@ -0,0 +1,52 @@
+    Others = 'others'
+}
+
+const cellTypeColors: ColorMap = {
```
I think your code here is overly complicated. Why not create an object to represent a celltype, which would include colors...etc, and instanciate a simple js object with all the combinations? I'm not sure I see the advantage of having the singleton and the lazy init, what you are doing with spliting between the celltypecolor, the celltypemap is splitting the inner components of an OO object, while the strategy here is simulating the behavior, more or less, of an object, trying to compose again values that are supposed to go together. 

If we push a little bit further, we could have simply a js object that would have this head:

```js
enum CellType {
	...
}

interface CellDesign = {
	type: CellType,
	color: string
}

const cellsConfig = {
  i: {type: CellType.Inter, color: '#FF887A'},
  b: {type: CellType.Muscle, color: ....},
... // etc    
}
```

Another option, as we don't have an at run time extensible set of cell types, you can have a superclass and subclasses for each types. 

For the strategy, I feel the singleton is unecessary here and it makes even less sense also considering that it already accesses `const` object that is declared as global. Usually using singleton is always controversial, and the point of the singleton is to avoid a global state, which is not really the case here in a way. The fact of using an enum to represent the two strategy and switching from one to the other using a function and relating the enum value to one or the other singleton looks like a misinterpretation of the strategy pattern (or weird implementation), but I understand the constraint as the menu needs to select one of the options.

aranega commented on applications/visualizer/frontend/src/helpers/twoD/coloringStrategy/ColoringStrategy.ts:
```diff
@@ -0,0 +1,27 @@
+    NEUROTRANSMITTER = 'Neurotransmitter',
+}
+
+export const getColoringStrategy = (option: ColoringOptions) => {
```
A simple object with the options would avoid the ifs:

```js

const colorThemes = {
	ColoringOptions.CELL_TYPE: ...,
	ColoringOptions.NEUROTRANSMITTER: ...
}

const getColoringStrategy = (option: ColoringOptions) => {
	return colorThemes[options]
}

aranega commented on applications/visualizer/frontend/src/models/workspace.ts:
```diff
@@ -117,4 +132,44 @@ export class Workspace {
+                draft.availableNeurons = {};
+
+                // Populate neuronsAvailable with neurons from all active datasets
+                neuronArrays.forEach(neurons => {
```
I wonder if you can flat and put in a set before the `forEach`, otherwise, you might treat the same neurons many times.

aranega commented on applications/visualizer/frontend/src/components/viewers/TwoD/TwoDMenu.tsx:
```diff
@@ -0,0 +1,246 @@
+    const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
+
+    const onZoomIn = () => {
+        if (cyRef.current) {
```
Is it possible to have only `cyRef` instead of `cyRef.current`? It feels strange to receive as parameter of a component a kind of low-level implementation object. This would make sense if the relation `cyRef.current` would be changed in the component (and it would be also weird to have a side-effect in the component).

aranega commented:
@afonsobspinto Thanks a lot for the PR! I made multiple small reviews and all the modifications are more related to details (let say) than really about the behavior of the PR in itself. Sorry for the review without message, but I'm still in the process of learning an extension Dario suggested to use for github (which is awesome, but which changes some buttons and behaviors).

afonsobspinto committed:
CELE-32 chore: Merge with develop

Salam-Dalloul committed:
#32 Style the 3d viewer component according to the theme defined by the figma design
- implement the left sidebar with the required buttons
#64 Style the connectome component according to the theme defined by the figma design
- implement the left sidebar with the required buttons

aranega commented:
Hi @Salam-Dalloul , thanks a lot for the modifications! I can see that the build is failing, could you take a look?

@afonsobspinto it seems that there is some conflicts on the PR, I'm not sure what's the state right now, but could you check when you have a chance? 

Thanks a lot :pray:

Salam-Dalloul committed:
fix eslint issue

Salam-Dalloul committed:
implement the dropdown to select the dataset to load/display

Salam-Dalloul committed:
update the twoDStyle

Salam-Dalloul committed:
add missing viewer icons + add tooltips + 3D viewer settings popover

afonsobspinto committed:
CELE-32 chore: Move main logic out of condition

afonsobspinto committed:
CELE-32 chore: Remove deprecated model

afonsobspinto committed:
CELE-32 chore: Remove typing import

afonsobspinto committed:
CELE-32 chore: Replace reference with current value

afonsobspinto committed:
CELE-32 chore: Move main logic out of condition

afonsobspinto committed:
CELE-32 feat: Simplify coloring strategies logic

afonsobspinto committed:
CELE-32 fix: Deduplicate neurons

afonsobspinto committed:
CELE-32 feat: Update get datasets route to include query parameters

afonsobspinto committed:
CELE-32 merge: Merge with develop

afonsobspinto committed:
CELE-32 style: Fix linting errors

afonsobspinto committed:
CELE-32 style: Fix linting errors backend

afonsobspinto committed:
CELE-32 style: Fix frontend build

aranega reviewed (commented):
Thanks a lot for the PR 🙏! All the code looks good to me, I'll test it quickly tomorrow and I'll merge it quickly 👌

afonsobspinto committed:
CELE-32 merge: Merge with CELE-32

afonsobspinto committed:
CELE-32 style: Apply linting rules

aranega committed:
CELE-63 Move some interfaces to the internal model

aranega committed:
CELE-63 Autocomplete only active datasets of a workspace

aranega committed:
Merge pull request #10 from MetaCell/feature/CELE-63

implement the dropdown to select the dataset to load/display
